### PR TITLE
feat: add EntryChunkPlugin to handle shebang and shims

### PR DESCRIPTION
**Pull request by fi3ework**
## Summary

Fix https://github.com/web-infra-dev/rslib/issues/391.
Fix https://github.com/web-infra-dev/rslib/issues/381.
Fix https://github.com/web-infra-dev/rslib/issues/166.
Fix https://github.com/web-infra-dev/rslib/issues/425.

### Overview

- Change the implementation of shim for `import.meta.url`. Previously, we can't use BannerPlugin to insert it into the top of the bundle as there might be a existing `"use strict"`, inject to the end also does not help as there will be a TDZ. In thi PR, using `processAssets` hook to insert the shim code to the correct position while preserve the leading `"use strict"`.
- Preserve the leading shebang directive for chunk entry modules, the implementation refers https://github.com/shuangwhywhy/webpack-shebang-plugin.
- Preserve the leading [React directive](https://react.dev/reference/rsc/directives) for chunk entry modules.

## Checklist

- [x] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/plugin-dts/src/dts.ts [hunk 1]
```diff
@@ -189,14 +189,12 @@ export async function generateDts(data: DtsGenOptions): Promise<void> {
...
-    entry = join(
-      declarationDir!,
-      relativePath,
-      basename(entrySourcePath),
-    ).replace(
-      /\.(js|mjs|jsx|ts|mts|tsx|cjs|cts|cjsx|ctsx|mjsx|mtsx)$/,
-      '.d.ts',
-    );
+    entry = join(declarationDir!, relativePath, basename(entrySourcePath))
+      .replace(/\?.*$/, '')
```
Comment (Timeless0911) [inline packages/plugin-dts/src/dts.ts:193]: leave a comment to specify to query syntax

File: tests/scripts/shared.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import assert from 'node:assert';
 import fs from 'node:fs';
-import { dirname, join, normalize } from 'node:path';
+import path, { dirname, join, normalize } from 'node:path';
```
Comment (Timeless0911) [inline tests/scripts/shared.ts:3]: why not use `{ basename }`

File: packages/core/src/config.ts [hunk 3]
```diff
@@ -1100,6 +1123,10 @@ async function composeLibRsbuildConfig(config: LibConfig, configPath: string) {
     cssModulesAuto,
   );
   const cssConfig = composeCssConfig(lcp, config.bundle);
+  const entryChunkConfig = composeEntryChunkConfig({
```
[the remark this answers is not in the document]
Comment (fi3ework) [reply packages/core/src/config.ts:1147]: Tweaked the return value, do not need to use `format` alongside now.

File: tests/scripts/shared.ts (hunk 2 above)
Comment (fi3ework) [reply tests/scripts/shared.ts]: I was intended to avoid the name conflict.

File: packages/core/src/config.ts [hunk 4]
```diff
@@ -600,10 +600,14 @@ const composeFormatConfig = ({
   }
 };
 
+type DeepRequired<T> = Required<{
```
Comment (Timeless0911) [inline packages/core/src/config.ts:603]: export in types or utils folder

File: packages/core/src/config.ts [hunk 5]
```diff
@@ -1020,7 +1041,7 @@ const composeTargetConfig = (
 const composeExternalHelpersConfig = (
   externalHelpers: boolean,
   pkgJson?: PkgJson,
-): RsbuildConfig => {
+) => {
```
Comment (Timeless0911) [inline packages/core/src/config.ts:1044]: why isolated declaration not throw an error

Comment (fi3ework) [reply packages/core/src/config.ts:1044]: Cause it's not exported?

**Review by Timeless0911 — APPROVED** [review]

**Pull request merged**